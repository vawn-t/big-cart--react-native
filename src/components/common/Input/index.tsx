import { useCallback, useState, memo, forwardRef, LegacyRef } from 'react';
import { TextInput, TouchableOpacity, View } from 'react-native';

// Icons
import { EyeIcon, EyeSlashIcon } from '@components/Icon';

// Types
import { IconName } from '@interfaces/index';

// Utils
import { getIcon } from '@utils/index';

// Styles
import styles from './styles';

interface IProps extends React.ComponentProps<typeof TextInput> {
  placeholder: string;
  icon: IconName;
  autoComplete?: 'email' | 'password' | 'off';
  containerStyles?: { [name: string]: number | string };
  isError?: boolean;
}

const Input = forwardRef(
  (
    {
      placeholder,
      autoComplete = 'off',
      containerStyles,
      icon,
      isError = false,
      style,
      ...props
    }: IProps,
    ref: LegacyRef<TextInput>,
  ) => {
    const [isSecured, setIsSecured] = useState(icon === IconName.Password);

    const IconComponent = getIcon(icon);

    const handleEyeToggle = useCallback(() => {
      setIsSecured((prev) => !prev);
    }, []);

    return (
      <View
        style={[
          styles.container,
          containerStyles,
          isError ? styles.error : null,
        ]}
        testID="Input"
      >
        <View style={styles.iconWrapper}>
          {IconComponent && <IconComponent />}
        </View>
        <TextInput
          ref={ref}
          testID="Input-TextInput"
          autoComplete={autoComplete}
          placeholder={placeholder}
          secureTextEntry={isSecured}
          style={[styles.baseInput, style]}
          {...props}
        />

        <TouchableOpacity onPress={handleEyeToggle} style={styles.iconWrapper}>
          {icon === IconName.Password && (
            <>
              {isSecured && <EyeIcon />}
              {!isSecured && <EyeSlashIcon />}
            </>
          )}
        </TouchableOpacity>
      </View>
    );
  },
);

Input.displayName = 'Input';

export default memo(Input);
