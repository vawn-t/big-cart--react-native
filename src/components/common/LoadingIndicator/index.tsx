import { ActivityIndicator, View } from 'react-native';
import { colors } from '@themes/index';

// Styles
import styles from './styles';

const LoadingIndicator = () => (
  <View style={styles.container}>
    <ActivityIndicator
      size="small"
      color={colors.primary.main}
      testID="loading-indicator"
    />
  </View>
);

export default LoadingIndicator;
