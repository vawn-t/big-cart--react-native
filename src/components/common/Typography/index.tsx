import { memo } from 'react';
import { Text } from 'react-native';

// Types
import { FontWeight, SizeType, TypoVariant } from '@interfaces/index';

// Styles
import styles from './styles';

interface IProps extends React.ComponentProps<typeof Text> {
  size?: SizeType;
  text: string | number;
  variant?: TypoVariant;
  fontWeight?: FontWeight;
}

const Typography = ({
  text,
  size = SizeType.Default,
  variant = TypoVariant.Paragraph1,
  fontWeight = FontWeight.Regular,
  ...props
}: IProps) => {
  const { style, ...rest } = props;

  return (
    <Text
      style={[styles[variant], styles[fontWeight], { fontSize: size }, style]}
      {...rest}
    >
      {text}
    </Text>
  );
};

export default memo(Typography);
