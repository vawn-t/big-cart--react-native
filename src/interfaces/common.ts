type ById<T> = {
  [id: number]: T;
};

export type { ById };
