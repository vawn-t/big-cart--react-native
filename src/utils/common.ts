// Types
import { ById } from '@interfaces/index';

export const removeById = <T>(byId: ById<T>, id: number) => {
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const { [id]: omitted, ...rest } = byId;
  return rest;
};

export const createById = <T extends { id: number }>(arr: T[]) => {
  return arr.reduce((obj, item) => Object.assign(obj, { [item.id]: item }), {});
};
