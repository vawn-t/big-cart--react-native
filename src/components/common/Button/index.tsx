import { memo } from 'react';
import { TouchableHighlight, View } from 'react-native';

// Themes
import { colors } from '@themes/index';

// Utils
import { getIcon } from '@utils/index';

// Types
import { FontWeight, IconName, TypoVariant } from '@interfaces/index';

// Components
import Typography from '../Typography';

// Styles
import styles from './styles';

interface IProps extends React.ComponentProps<typeof TouchableHighlight> {
  title: string;
  icon?: IconName;
  onPress: () => void;
}

const Button = ({
  onPress,
  icon,
  title,
  style,
  disabled,
  ...props
}: IProps) => {
  const Icon = icon && getIcon(icon);

  return (
    <TouchableHighlight
      style={[styles.container, disabled ? styles.disabledButton : null, style]}
      onPress={onPress}
      underlayColor={colors.background.light}
      disabled={disabled}
      {...props}
    >
      <View style={styles.contentWrapper}>
        {Icon && (
          <View style={styles.iconWrapper}>
            <Icon testID="Button-icon" />
          </View>
        )}
        <Typography
          text={title}
          variant={TypoVariant.Paragraph2}
          style={styles.baseTitle}
          key={title}
          fontWeight={FontWeight.SemiBold}
        />
      </View>
    </TouchableHighlight>
  );
};

export default memo(Button);
