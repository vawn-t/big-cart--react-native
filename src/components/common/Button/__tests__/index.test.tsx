import { fireEvent, render, screen } from '@utils/index';
import { Button } from '@components/common';
import { IconName } from '@interfaces/index';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock'),
);

const ownProps: React.ComponentProps<typeof Button> = {
  onPress: jest.fn(),
  title: 'Hello world',
  icon: IconName.Password,
};

const setup = (overrideProp = {}) => {
  const props = { ...ownProps, ...overrideProp };

  return render(<Button {...props} />);
};

describe('Button component', () => {
  it('renders correctly', () => {
    const tree = setup();
    expect(tree).toMatchSnapshot();
  });

  it('onPress should be called', () => {
    setup();
    fireEvent.press(screen.getByText(ownProps.title));

    expect(ownProps.onPress).toHaveBeenCalled();
  });

  it('should render icon', () => {
    setup({ icon: IconName.User });

    expect(screen.getByTestId('Button-icon')).toBeTruthy();
  });
});
