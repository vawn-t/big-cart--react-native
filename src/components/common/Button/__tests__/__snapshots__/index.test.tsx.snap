// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component renders correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      gestureResponseDistance={
        {
          "bottom": -1,
          "end": -1,
          "start": -1,
          "top": -1,
        }
      }
      nativeBackButtonDismissalEnabled={false}
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onHeaderBackButtonClicked={[Function]}
      onNativeDismissCancelled={[Function]}
      onTransitionProgress={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      sheetAllowedDetents="large"
      sheetCornerRadius={-1}
      sheetExpandsWhenScrolledToEdge={true}
      sheetGrabberVisible={false}
      sheetLargestUndimmedDetent="all"
      stackPresentation="push"
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      swipeDirection="horizontal"
    >
      <View
        accessibilityElementsHidden={false}
        importantForAccessibility="auto"
        style={
          {
            "flex": 1,
            "flexDirection": "column-reverse",
          }
        }
      >
        <View
          collapsable={false}
          style={
            [
              {
                "flex": 1,
              },
              {
                "backgroundColor": "rgb(242, 242, 242)",
              },
              undefined,
            ]
          }
        >
          <View
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              [
                {
                  "alignItems": "center",
                  "backgroundColor": "#6CC51D",
                  "borderRadius": 5,
                  "flexDirection": "row",
                },
                null,
                undefined,
              ]
            }
          >
            <View
              style={
                {
                  "flexGrow": 1,
                  "paddingVertical": 20,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "flex-start",
                    "bottom": 0,
                    "justifyContent": "center",
                    "left": 30,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                  }
                }
              >
                <RNSVGSvgView
                  bbHeight="23"
                  bbWidth="18"
                  fill="none"
                  focusable={false}
                  height={23}
                  style={
                    [
                      {
                        "backgroundColor": "transparent",
                        "borderWidth": 0,
                      },
                      {
                        "flex": 0,
                        "height": 23,
                        "width": 18,
                      },
                    ]
                  }
                  testID="Button-icon"
                  width={18}
                >
                  <RNSVGGroup
                    fill={null}
                    propList={
                      [
                        "fill",
                      ]
                    }
                  >
                    <RNSVGPath
                      d="M15.094 23H2.156A2.159 2.159 0 0 1 0 20.844V10.78c0-1.19.968-2.156 2.156-2.156h12.938c1.188 0 2.156.967 2.156 2.156v10.063c0 1.19-.968 2.156-2.156 2.156ZM2.156 10.062a.72.72 0 0 0-.719.72v10.062a.72.72 0 0 0 .72.718h12.937a.72.72 0 0 0 .719-.718V10.78a.72.72 0 0 0-.72-.719H2.157Z"
                      fill={
                        {
                          "payload": 4287006857,
                          "type": 0,
                        }
                      }
                      propList={
                        [
                          "fill",
                        ]
                      }
                    />
                    <RNSVGPath
                      d="M13.656 10.063a.719.719 0 0 1-.719-.72V5.75a4.318 4.318 0 0 0-4.312-4.313A4.318 4.318 0 0 0 4.312 5.75v3.594a.719.719 0 0 1-1.437 0V5.75A5.756 5.756 0 0 1 8.625 0a5.756 5.756 0 0 1 5.75 5.75v3.594a.719.719 0 0 1-.719.719Z"
                      fill={
                        {
                          "payload": 4287006857,
                          "type": 0,
                        }
                      }
                      propList={
                        [
                          "fill",
                        ]
                      }
                    />
                  </RNSVGGroup>
                </RNSVGSvgView>
              </View>
              <Text
                style={
                  [
                    {
                      "color": "#000000",
                    },
                    {
                      "fontFamily": "Poppins-SemiBold",
                    },
                    {
                      "fontSize": 15,
                    },
                    {
                      "color": "#FFFFFF",
                      "textAlign": "center",
                    },
                  ]
                }
              >
                Hello world
              </Text>
            </View>
          </View>
        </View>
      </View>
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backTitleVisible={true}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        largeTitleHideShadow={false}
        title="test"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
